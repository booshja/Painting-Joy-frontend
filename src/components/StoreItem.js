// dependencies
import React from "react";
import styled from "styled-components";
import { useHistory } from "react-router";

const StyledDiv = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    border-radius: 4px;
    box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
    height: 200px;
    width: 165px;
    padding: 1rem;
    margin-bottom: 1rem;
    position: relative;
    cursor: pointer;
`;

const StyledImg = styled.img`
    height: 133px;
    width: 110px;
`;

const StyledOverlay = styled.div`
    height: 26px;
    width: 110px;
    background-color: rgba(0, 0, 0, 0.5);
    color: #ffffff;
    font-family: "News Cycle", sans-serif;
    font-size: 1.2rem;
    letter-spacing: 1px;
    text-transform: uppercase;
    display: flex;
    justify-content: center;
    align-items: center;
    position: absolute;
    top: 4rem;
    left: 1.7rem;
`;

const StyledTitle = styled.h3`
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    text-align: center;
`;

const StyledPrice = styled.p`
    font-family: "News Cycle", sans-serif;
    font-size: 1rem;
    letter-spacing: 1px;
    text-align: center;
    padding-top: 0.25rem;
`;

const StoreItem = ({ item }) => {
    const history = useHistory();

    const handleClick = (e) => {
        history.push(`/store/item/${e.target.id.slice(2)}`);
    };

    return (
        <StyledDiv id={"d-" + item.id} onClick={(e) => handleClick(e)}>
            <StyledImg
                src={
                    process.env.REACT_APP_BACKEND_URL +
                    `items/item/${item.id}/image`
                }
                alt={item.title}
                id={"i-" + item.id}
            />
            {item.isSold && (
                <StyledOverlay id={"o-" + item.id}>Sold</StyledOverlay>
            )}
            <StyledTitle id={"t-" + item.id}>{item.title}</StyledTitle>
            {item.isSold ? (
                <StyledPrice id={"p-" + item.id}>${item.price}</StyledPrice>
            ) : (
                <StyledPrice id={"p-" + item.id}>
                    ${item.price} + ${item.shipping} shipping
                </StyledPrice>
            )}
        </StyledDiv>
    );
};

export default StoreItem;
