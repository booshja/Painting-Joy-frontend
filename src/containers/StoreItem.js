// dependencies
import React, { useEffect, useState } from "react";
import styled from "styled-components";
import axios from "axios";
import { useParams } from "react-router-dom";
// components
import { GoBack } from "../components";

const StyledStoreItem = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0 5%;
`;

const StyledImg = styled.img`
    width: 100%;
`;

const StyledTitle = styled.h1`
    font-family: "News Cycle", sans-serif;
    font-size: 1.8rem;
    letter-spacing: 1px;
    font-weight: 700;
    align-self: flex-start;
    margin: 1rem 0;
`;

const StyledText = styled.p`
    font-family: "News Cycle", sans-serif;
    font-size: 1.4rem;
    letter-spacing: 1px;
    align-self: flex-start;
    text-align: left;
    line-height: 2rem;
    margin-bottom: 1rem;
    width: 90%;
`;

const StyledButton = styled.button`
    background-color: #207070;
    border: 2px solid #207070;
    border-radius: 4px;
    color: #ffffff;
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.2rem;
    letter-spacing: 1px;
    padding: 2.5px 10px;
    align-self: flex-start;

    &:hover {
        color: #207070;
        background-color: #ffffff;
    }
`;

const StoreItem = () => {
    const [isLoading, setIsLoading] = useState(true);
    const [item, setItem] = useState({});
    const { itemId } = useParams();

    useEffect(() => {
        async function getItem() {
            const res = await axios(
                process.env.REACT_APP_BACKEND_URL + `items/item/${itemId}`
            );
            setItem(() => res.data.item);
        }
        getItem();
        setIsLoading(false);
    }, []);

    const handleClick = () => {
        // TODO - add item to cart
        // TODO - need to add state for cart in App
        // TODO - need to add cart to mobile menu
    };

    if (isLoading) {
        return <p>Loading...</p>;
    }

    return (
        <StyledStoreItem>
            <GoBack to={"/store"} />
            <StyledImg
                src={
                    process.env.REACT_APP_BACKEND_URL +
                    `items/item/${itemId}/image`
                }
                alt={item.title}
            />
            <StyledTitle>{item.name}</StyledTitle>
            <StyledText>{item.description}</StyledText>
            <StyledText>
                ${item.price} + ${item.shipping} shipping
            </StyledText>
            <StyledButton onClick={handleClick}>Buy Me!</StyledButton>
        </StyledStoreItem>
    );
};

export default StoreItem;
