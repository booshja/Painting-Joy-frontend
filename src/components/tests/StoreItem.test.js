import { render } from "@testing-library/react";
import StoreItem from "../StoreItem";

const item = {
    id: 0,
    name: "Item name",
    price: "99.99",
    shipping: "100.05",
    isSold: false,
};

const soldItem = {
    id: 0,
    name: "Item name",
    price: "99.99",
    shipping: "100.05",
    isSold: true,
};

/** Smoke Test */
it("renders without crashing, isSold false", () => {
    render(<StoreItem item={item} />);
});

it("renders without crashing, isSold true", () => {
    render(<StoreItem item={soldItem} />);
});

/** Snapshot Test */
it("matches snapshot, isSold false", () => {
    const { asFragment } = render(<StoreItem item={item} />);
    expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot, isSold true", () => {
    const { asFragment } = render(<StoreItem item={soldItem} />);
    expect(asFragment()).toMatchSnapshot();
});

/** RTL Tests */
it("renders correctly, isSold false", () => {
    const { getByAltText, getByText } = render(<StoreItem item={item} />);

    getByAltText("Item name");
    getByText("Item name");
    getByText("$99.99 + $100.05 shipping");
});

it("renders correctly, isSold true", () => {
    const { getByAltText, getByText } = render(<StoreItem item={soldItem} />);

    getByAltText("Item name");
    getByText("Sold");
    getByText("Item name");
});
