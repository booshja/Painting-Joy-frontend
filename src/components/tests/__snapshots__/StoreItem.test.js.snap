// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot, isSold false 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM jnsyBH"
    id="d-0"
  >
    <img
      alt="Item name"
      class="sc-gtsrHT hCRpdD"
      id="i-0"
      src="http://localhost:5000/items/item/0/image"
    />
    <h3
      class="sc-hKFxyN csRcPi"
      id="t-0"
    >
      Item name
    </h3>
    <p
      class="sc-eCApnc hRhdvA"
      id="p-0"
    >
      $99.99 + $100.05 shipping
    </p>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot, isSold true 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM jnsyBH"
    id="d-0"
  >
    <img
      alt="Item name"
      class="sc-gtsrHT hCRpdD"
      id="i-0"
      src="http://localhost:5000/items/item/0/image"
    />
    <div
      class="sc-dlnjwi eYQnHe"
      id="o-0"
    >
      Sold
    </div>
    <h3
      class="sc-hKFxyN csRcPi"
      id="t-0"
    >
      Item name
    </h3>
    <p
      class="sc-eCApnc hRhdvA"
      id="p-0"
    >
      $99.99
    </p>
  </div>
</DocumentFragment>
`;
